src/StrongPath.hs

```diff
@@ -821,16 +845,89 @@ castFile _ = impossible
+      quotePat = err "pattern",
+      quoteType = err "type",
+      quoteDec = err "declaration"
+    }
+  where
+    err what x = fail ("unexpected " ++ what ++ ", must be expression: " ++ x)
+
+-- TODO: This means that custom types provided by users to describe the Files/Dirs also need to be Typeable? So, users can't do just `data MyFile` anymore, instead they have to do `data MyFile deriving Typeable`?
+instance forall s b t. (Typeable s, Typeable b, Typeable t) => TH.Lift (Path s b t) where
+  lift sp = do
+    let s = TH.ConT $ getTCName (Proxy :: Proxy s)
+        b = TH.ConT $ getTCName (Proxy :: Proxy b)
+        t = TH.ConT $ getTCName (Proxy :: Proxy t)
+    case sp of
+      -- TODO: Could we use Typeable here to figure out data constructor and its args and use
+      --   that to construct the final result, so we don't have to pattern match?
+      RelDir path relPrefix -> [|RelDir $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      RelFile path relPrefix -> [|RelFile $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsDir path -> [|AbsDir $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsFile path -> [|AbsFile $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+      RelDirP path relPrefix -> [|RelDirP $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      RelFileP path relPrefix -> [|RelFileP $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsDirP path -> [|AbsDirP $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsFileP path -> [|AbsFileP $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+      RelDirW path relPrefix -> [|RelDirW $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      RelFileW path relPrefix -> [|RelFileW $(lift path) $(lift relPrefix) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsDirW path -> [|AbsDirW $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+      AbsFileW path -> [|AbsFileW $(lift path) :: Path $(pure s) $(pure b) $(pure t)|]
+    where
+      getTCName :: Typeable a => proxy a -> TH.Name
+      getTCName a = TH.Name occ flav
+        where
+          tc = D.typeRepTyCon (D.typeRep a)
+          occ = TH.OccName (D.tyConName tc)
+          flav =
+            TH.NameG
+              TH.TcClsName
+              (TH.PkgName (D.tyConPackage tc))
+              (TH.ModName (D.tyConModule tc))
+
+absdir, absdirP, absdirW :: QuasiQuoter
+absdir = qq parseAbsDir
+absdirP = qq parseAbsDirP
+absdirW = qq parseAbsDirW
+
+absfile, absfileP, absfileW :: QuasiQuoter
+absfile = qq parseAbsFile
+absfileP = qq parseAbsFileP
+absfileW = qq parseAbsFileW
+
+reldir, reldirP, reldirW :: QuasiQuoter
+reldir = qq parseRelDir
+reldirP = qq parseRelDirP
+reldirW = qq parseRelDirW
+
+relfile, relfileP, relfileW :: QuasiQuoter
+relfile = qq parseRelFile
+relfileP = qq parseRelFileP
+relfileW = qq parseRelFileW
```

Comment:
These are throwing compile time error. It seems that `qq` needs to know that something is `Typeable`, probably type that user provides as part of Path to describe dir/file, but it can't know that for sure? I am not sure how to think about this hm.